core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageOutputStream.java

```diff
@@ -0,0 +1,160 @@
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamMessageOutputStream<T> extends OutputStream {
+
+    private final StreamMessage<T> source;
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private final ByteBuf buffer = Unpooled.buffer();
+    private boolean transferred;
+    private boolean closed;
+
+    StreamMessageOutputStream(StreamMessage<T> source,
+                              StreamWriter<T> writer,
+                              Function<? super HttpData, ? extends T> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        this.source = source;
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        buffer.writeBytes(b, off, len);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        ensureTransferred();
+        drain();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        flush();
+        closed = true;
+        writer.close();
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+        if (!writer.isOpen()) {
+            closed = true;
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureTransferred() {
+        if (!transferred) {
+            transferred = true;
+            final StreamMessageTransferSubscriber<T> subscriber = new StreamMessageTransferSubscriber<>(writer);
+            source.subscribe(subscriber);
+            subscriber.whenComplete.exceptionally(e -> {
+                closed = true;
+                writer.close();
+                return Exceptions.throwUnsafely(e);
+            }).join();
+        }
+    }
+
+    private void drain() {
+        if (!buffer.isReadable() || !writer.isOpen()) {
+            return;
+        }
+        final int readableBytes = buffer.readableBytes();
+        final byte[] data = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            data[i] = buffer.readByte();
+        }
```

Comment:
How about using an unpooled heap `ByteBuf` so it is always backed by a byte array? Then, we could use `ByteBuf.array()` to access the underlying array and copy it into a new `HttpData` using `HttpData.copyOf(byte[],...)`.